### ISSUE TITLE
Parameter truncation when executing SQL statement using `fbdb` library

### ISSUE THREAD

**Issue by cwangfr** [hipercompl/fbdb#4]
**Issue: Parameter truncation when executing SQL statement using `fbdb` library**

**Description:**
While using the `fbdb` library for database operations, I encountered an issue where parameters are not passed completely in an `INSERT` SQL statement. Specifically, when passing a `VARCHAR` field, the value seems to get truncated, causing a foreign key constraint violation in the database.

**Steps to Reproduce:**
1. Prepare an `INSERT` statement with parameters that include a `VARCHAR(32)`.
2. Execute the following code:
   
   ```dart
   debugPrint("UUID: ${uuid.v4().replaceAll("-", "")}");
   debugPrint("OrderDetailId: ${orderDetail.orderDetailId}");
   debugPrint("OrderCount: ${orderDetail.orderCount}");
   debugPrint("FinishedAt: ${eponch.value}");

   await q.execute(
       sql: "INSERT INTO ORDER_DETAIL_LOG(ID, DET_ID, DELV_COUNT, FINISHED_AT) VALUES (?,?,?,?)",
       parameters: [
           uuid.v4().replaceAll("-", ""), // generate a UUID
           orderDetail.orderDetailId,
           orderDetail.orderCount,
           eponch.value
       ]);
   ```

   Output from `debugPrint`:
   ```dart
   I/flutter (28350): UUID: 508599ecd0814f89a3ca1114bdd84b27
   I/flutter (28350): OrderDetailId: 3408108B67544334A99AA8FB237D4EE5
   I/flutter (28350): OrderCount: 1
   I/flutter (28350): FinishedAt: 1729088428
   ```

3. Observe the error in the database:

   ```dart
   I/flutter (28350): violation of FOREIGN KEY constraint "FK_ORDER_DETAIL_LOG" on table "ORDER_DETAIL_LOG"
   I/flutter (28350): -Foreign key reference target does not exist
   I/flutter (28350): -Problematic key value is ("DET_ID" = '3408108B67544334A99AA8FB237D4E')
   ```

**Expected Behavior:**
The `orderDetailId` (which in this case is a 32-character string) should be passed in full to the database, and the foreign key constraint should not fail since the referenced key exists in the target table.

**Actual Behavior:**
The `orderDetailId` is getting truncated to `3408108B67544334A99AA8FB237D4E` (missing 2 characters), causing the foreign key constraint to fail.

**Environment:**
- `fbdb` library version: 1.1.0
- Dart SDK version: 3.5.3
- Flutter version: 3.24.3
- Database: Firebird 2.5.8 Windows 64bit
- Runtime: Android Emulator API 30
  
**Additional Information:**
- The `DET_ID` field in the `ORDER_DETAIL_LOG` table is of type `VARCHAR(32)`, and I am passing a 32-character string from 'ORDER_DETAIL' table.

**Suspected Cause:**
It seems that when passing parameters to the SQL statement using `fbdb`, the string might be getting truncated or mishandled during binding or in the underlying database driver.

**Temporary Workaround:**
I haven't identified a solid workaround yet, but the issue seems to be with parameter binding and passing, particularly with the UUID string being truncated.

**Request:**
Please investigate why parameters passed to SQL statements via `fbdb` are being truncated and provide guidance on how to correctly handle this or fix the truncation issue in the library.

**Comment by ttyrakow** [comment]
I can confirm it does happen on Android, but not on Windows nor Linux. The test case: [https://github.com/hipercompl/fbdb/blob/f5990785f6cdde184affab128ebe97fb520170ae/test/fbdb_query_test.dart#L316-L332](https://github.com/hipercompl/fbdb/blob/f5990785f6cdde184affab128ebe97fb520170ae/test/fbdb_query_test.dart#L316-L332) passes on both Windows and Linux (haven't tested on MacOS - no access to Apple hardware).
However, I've stitched together a simple Android Flutter application, doing more or less the same as the test case, and the last two characters are indeed truncated. I'll look into this. I'm not sure if it's a problem with fbdb, or with the Android version of the Firebird client library (or the joint of the two), I need some time to investigate.
Thank you for reporting the issue (and for the high quality of your report).

**Comment by ttyrakow** [comment]
Found, evaluated, fixed :wink:.
Uploaded version 1.1.1 to pub.dev.
I'd appreciate if you could update your fbdb dependency to 1.1.1 and confirm if the problem is gone.
Thanks again for reporting the issue, it was an interesting one.
Just for your information (you made a significant effort to report the issue properly, so in return I'll make some effort to explain to you what actually happened): the metadata (received from the Firebird server) reported a parameter of size 32 (bytes). A `VARCHAR` field / parameter is supposed to have a 2-byte character count written at the beginning of its data buffer. I've counted those 2 bytes as being part of the parameter length (i.e. the text plus the two bytes had to fit inside a 32-byte parameter), while it turned out not to be the case. The parameter length 32 in case of `VARCHAR` means "32 bytes **plus** 2-byte length", and not "32 bytes **including** 2-byte length". That's why the last 2 characters were chopped off (the 2-byte character count took the space of 2 characters, and fbdb truncated the remaining text not to overrun the buffer). What's even more interesting is why the tests didn't catch that (even a specific test, which I've added just to test this very issue). The databases used by unit tests are UTF-8 encoded. In that case, the parameter size for `CHAR`/`VARCHAR` types is actually 4 times larger than the length of the database field counted in characters (as each Unicode code point can take up to 4 bytes when encoded as UTF-8). Since the actual texts used in test cases don't require 4 bytes for each character (they're mostly plain ASCII strings), there was plenty of room left for the 2-byte character count and fbdb saw no need to truncate the text. 
That's more or less the full story. The fault was mine, I wasn't imaginative enough to provide for such cases, but fortunately it's now resolved and the fix was quick and easy (finding the cause was much harder, though).

**Comment by cwangfr** [comment]
Hi,

Thank you for the quick fix and the detailed explanation!

I’ve updated fbdb to version 1.1.1 and tested it, and I can confirm that the issue is now resolved. Everything works perfectly, and the parameter truncation problem is gone.

I really appreciate the effort you put into explaining the root cause of the issue and how it was fixed. It’s always great to gain a deeper understanding of what’s going on under the hood. Thanks again for your hard work and the fast turnaround!

Best regards

**Comment by ttyrakow** [comment]
Fixed in version 1.1.1.
